src/components/Button/index.js

```diff
@@ -6,20 +6,49 @@ const Button = ({
   text,
   buttonStyle,
   isDisabled = false,
-  className,
+  className = '',
   children,
   type = 'button',
+  onClick,
   ...props
 }) => {
+  const loadingRef = React.useRef(false);
+  const promiseRef = React.useRef();
+
+  const buttonClick = () => {
```

Comment:
can this be simplified with `async`?

```js
const [loading, setLoading] = React.useState(false);
const buttonClick = async () => {
  setLoading(true);
  await onClick();
  setLoading(false)
}
```

Reply:
Keeping track of the promise just helps prevent multiple clicks from triggering async actions until the previous one has resolved and I think we would have to move ```setLoading``` into a ```React.useEffect```

If there is a ```Promise.all``` polyfill we could do this

```javascript
  const loadingRef = React.useRef(false);
  const promiseRef = React.useRef();
  const updateRefs = (isLoading = false, promise = null) => {
    loadingRef.current = isLoading;
    promiseRef.current = promise;
  }

  const buttonClick = () => {
    if (typeof onClick !== 'function') return onClick;
    if (loadingRef.current) return promiseRef.curent;

    const result = onClick();
    updateRefs(true, result);
    Promise.all([result]).then(() => updateRefs());
    return result;
  }
```

If not...
Do you think this is clearer?

```javascript
  const loadingRef = React.useRef(false);
  const promiseRef = React.useRef();
  const updateRefs = (isLoading = false, promise = null) => {
    loadingRef.current = isLoading;
    promiseRef.current = promise;
  }

  const buttonClick = () => {
    if (typeof onClick !== 'function') return onClick;
    if (loadingRef.current) return promiseRef.curent;

    const result = onClick();
    const isPromise = result && typeof result.then === 'function';

    if (isPromise) {
      updateRefs(true, result);
      result.then(() => updateRefs());
    }
    return result;
  }
```

Reply:
hmm but what if we only execute `onClick` when `loading===false`? That way we ensure that async actions are only triggered once at a time.

```
const [loading, setLoading] = React.useState(false);
const buttonClick = async () => {
  if (loading) return;
  setLoading(true);
  await onClick();
  setLoading(false)
}
```

I assumed no clicks would be executed since the button gets disabled while a promise is being resolved. Does this works?

Reply:
Yeah I think this would work. 

Some implications though worth testing/considering...
- If the ```onClick``` event did something that took a long time to compute would it display a loading state right? 
That does seem to have advantages over just checking for a promise. For instance, if you were doing something computationally expensive on the front end?
- We might need to put it in a ```useEffect``` incase it set state after the button was unmounted? Or maybe just keep it as a ```useRef```

Reply:
> If the onClick event did something that took a long time to compute would it display a loading state right?

Yes, `loading` should be `true` while the promise is being executed.

> We might need to put it in a useEffect incase it set state after the button was unmounted?

Yes, that seems like a good solution. Good catch!